Resumable file uploads with tus.io

mvdbeek opened the pull request:
We can currently either use chunked uploads, or use the nginx upload module.
Both of these have significant disadvantages:

The chunked uploader POSTs chunks of files to /api/upload. This is a blocking operation (that can be a significant issue for slow simultaneous uploads), chunks are buffered to disk both in nginx and uwsgi (if used) and are re-assembled.
We're using a custom API to do this, and I think the only client that uses this protocol is the uploader in Galaxy's GUI.

We can also use nginx's upload module, but this module needs to be compiled at nginx build time, is not commonly distributed with the standard nginx package, isn't actively developed (as far as I know) and as far as I understand it, does not support chunked upload. There is in principle support for resumable uploads, but I don't think we've ever configured this.

I started implementing support for [nginx-big-upload](https://github.com/pgaertig/nginx-big-upload) which only requires the nginx lua module, but noticed that we'd still need to add custom headers (not a huge problem). Also it doesn't seem to be actively maintained or widely used.
With some more googling I found https://tus.io/, which sounds really good, especially regarding the implementations.

There's a WSGI middleware that I was able to plug directly into Galaxy, which we can use as our out-of-the-box-no-config-needed tus server. 

Of course the middleware doesn't solve the blocking issue, but in a production setup we canto set up a [tusd server](https://github.com/tus/tusd) and proxy requests to `/api/uploads/resumable_upload` to the tusd server. The server can send various hooks, so on complete it can send the the re-assembled file path to `/api/uploads/resumable_upload/{tmpfile}`.

In practice that works like this:
```
tusd -host localhost -port 1080 -upload-dir=<galaxy_root>/database/tmp -hooks-http=<galaxy_url>/api/upload/hooks -hooks-http-forward-headers=X-Api-Key,sessioncookie
```

Adjust ./database/tmp to new_file_path, and this should work. Now we just need to proxy `api/upload/resumable_upload` to `<tusd_port>/files`

If we don't want to include another server we can probably also use an [nginx lua block](https://github.com/mmatuska/lua-tus-server) to do the same.

The protocol (and our middleware) also explicitly supports checksums for the chunks.
~I haven't started work on the client yet, but~ I've included a small script in /scripts/resumable_upload.py in case anyone want to play with this (the async library is of course not necessary ...). This should also be a good basis for inclusion in client libraries such as bioblend that currently do not perform chunked uploads (IIRC).

I think (when everything is implemented) this should fix https://github.com/galaxyproject/galaxy/issues/5695, https://github.com/galaxyproject/galaxy/issues/5536 and https://github.com/galaxyproject/galaxy/issues/9819

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

mvdbeek committed:
Backend Prototype for resumable uploads via TUS

The middleware can be replaced by running a [tusd
server](https://github.com/tus/tusd) or
[nginx tus location](https://github.com/mmatuska/lua-tus-server) in
front of Galaxy.
Includes a sample script that can upload datasets into a Galaxy history.

mvdbeek committed:
Add simple API test for tus endpoint

mvdbeek committed:
Adjust endpoint so it can receive tusd hooks

Without any proxying and tusd running as a separate service:
``
tusd -upload-dir=./database/tmp -hooks-http=http://localhost:8000/api/upload/_resumable_upload -hooks-http-forward-headers=X-Api-Key --hooks-enabled-events=post-finish
```

Adjust ./database/tmp to new_file_path, and this should work. Now we
just need to proxy `api/upload/resumable_upload` to `<tusd_port>/files`
and we're good for a production setup.

mvdbeek committed:
Add admin documentation for running tusd + nginx

mvdbeek committed:
List second endpoint

mvdbeek committed:
Use my fork of tusfilter

mvdbeek committed:
Use non-async tus client library

mvdbeek committed:
Implement chunked upload using tus-js-client

mvdbeek committed:
Decrease chunk size, log time taken

mvdbeek committed:
Use client-logic to run upload tool

It is pretty cool to use hooks and an alternative tool submission
endpoint, but this is less invasive. We'd have to do this anyway for
cases where the upload is already on the server and the pre-finish hook
doesn't fire.

mvdbeek committed:
Update docs with correct headers to forward

mvdbeek committed:
Fix documented nginx location for tus

jmchilton commented on scripts/resumable_upload.py:
```diff
@@ -0,0 +1,62 @@
+    # Extract session from created upload URL
+    session_id = uploader.url.rsplit('/', 1)[1]
+    # This feels a bit more user-friendly ?
+    tool_id = 'upload1'
```
I'm a little sad you based this on upload1 and not the data fetch API. This seems backward facing to me.

mvdbeek replied:
I can fix that, I just looked at what the client does

mvdbeek replied:
Actually, do we have an example or a test for single files ?

mvdbeek replied:
Nevermind, found it

mvdbeek committed:
We don't need a session for chunked upload

mvdbeek committed:
Missing slash

mvdbeek committed:
Fix double-success

mvdbeek committed:
Drop chunksize

mvdbeek committed:
Use fetch api in api tests

mvdbeek committed:
Adapt scripts/resumable_upload.py to fetch API

Finding out that the `files` parameter is required took way more time
than I want to admit ... .
